src/alloc.rs

```diff
@@ -37,48 +37,22 @@ pub struct AllocWithInfo;
 unsafe impl GlobalAlloc for AllocWithInfo {
     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
         let ptr = libc::malloc(layout.size() as libc::size_t) as *mut u8;
-
-        ALLOC_LOCK.lock();
-        match ALLOC_INFO {
-            Some(ref mut pm) => pm.insert(ptr as usize, layout.size(), false),
-            None => {
-                let mut al = AllocList::new();
-                al.insert(ptr as usize, layout.size(), false);
-                ALLOC_INFO = Some(al);
-            }
-        };
-        ALLOC_LOCK.unlock();
+        AllocMetadata::insert(ptr as usize, layout.size(), false);
         ptr
     }
 
     unsafe fn dealloc(&self, ptr: *mut u8, _layout: Layout) {
         libc::free(ptr as *mut libc::c_void);
-        ALLOC_LOCK.lock();
-        ALLOC_INFO.as_mut().unwrap().remove(ptr as usize);
-        ALLOC_LOCK.unlock();
+        AllocMetadata::remove(ptr as usize);
     }
 
     unsafe fn realloc(&self, ptr: *mut u8, _layout: Layout, new_size: usize) -> *mut u8 {
         let new_ptr = libc::realloc(ptr as *mut libc::c_void, new_size) as *mut u8;
 
-        ALLOC_LOCK.lock();
-        if ptr.is_null() {
-            // realloc is equivalent to malloc if called with a null pointer,
-            // thus we must insert the new block info in the ALLOC_INFO list.
-            ALLOC_INFO
-                .as_mut()
-                .unwrap()
-                .insert(ptr as usize, new_size, false);
-        } else {
-            // In all other cases, the memory block is resized and a new ptr
-            // is returned. This may differ from the original ptr if the new
-            // block is too large to fit in the same space as the old block
-            ALLOC_INFO
-                .as_mut()
-                .unwrap()
-                .update(ptr as usize, new_ptr as usize, new_size);
-        }
-        ALLOC_LOCK.unlock();
+        debug_assert!(!ptr.is_null());
```

Comment:
I must admit, I'd be inclined to make this an `assert`, because it's such a strong condition that we need satisfied.

Reply:
Okay